### PR TITLE
Add support for cartopy projections with geographic data

### PR DESCRIPTION
**Pull request by munkm**
## PR Summary

This PR adds in support for projecting geographic data sources on various map types. As it stands, the code in this PR assumes that geographic data is read in as `PlateCarree` type and can be projected onto the maps listed here: https://scitools.org.uk/cartopy/docs/v0.15/crs/projections.html.  

## PR Checklist

- [x] Code passes flake8 checker
- [x] New features are documented, with docstrings and narrative docs
- [x] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by MatthewTurk** [comment]
[Cartopy projections for tomography stuff](https://trello.com/c/Z2yR2HZw/3-cartopy-projections-for-tomography-stuff)

**Comment by munkm** [comment]
Ok, I think this is pretty close to a working state. 

This branch:
* Changes the axes for a geographic coordinate system-based dataset to be an instance of `cartopy.mpl.geoaxes.GeoAxes`. These axes can then be annotated as you would with cartopy. 
* The default transform and projection type is assumed to be `PlateCarree` 
* includes a new function `set_mpl_projection()` which can be used to set the projection on a dataset that has a detected geographic coordinate system

a default slice plot with this is `PlateCarree`
```
ds = yt.load_uniform_grid(data, sizes, 1.0, geometry=("geographic", dims),
                          bbox=bbox)
p = yt.SlicePlot(ds, "altitude", 'AIRDENS')
p.show()
 ```

image

This can be annotated like any other plot with GeoAxes to give contextual information. 

```p = yt.SlicePlot(ds, "altitude", 'AIRDENS')
p._setup_plots()
p.plots['AIRDENS'].axes.set_global()
p.plots['AIRDENS'].axes.coastlines()
p.show() 
```

image

now `set_mpl_projection()` can be used to transform and project this data onto new axes

a few examples: 
```
p.set_mpl_projection('AlbersEqualArea')
p._setup_plots()
p.plots['AIRDENS'].axes.set_global()
p.plots['AIRDENS'].axes.coastlines()
p.show() 
``` 

image

and 

```
p.set_mpl_projection('InterruptedGoodeHomolosine')
p._setup_plots()
p.plots['AIRDENS'].axes.set_global()
p.plots['AIRDENS'].axes.coastlines()
p.show() 
```

image

**Comment by ngoldbaum** [comment]
ping @dopplershift and @pelson you guys might be interested in this :)

**Comment by ngoldbaum** [comment]
@zssherman how do you want to deal with upstreaming the `cf_radial` frontend? That should probably happen separately from this.

**Comment by ngoldbaum** [comment]
I think we probably want to add a way to call `set_global()` and `coastlines()` on the plot object itself, but that can come separately, I guess via the plot callback that @zssherman was working on.

This is really, *really* awesome.

**Comment by zssherman** [comment]
Nathan,
I can do it seperately if you would like? Sadly i've been out sick, so it might take a day or two.


Sent from my Verizon, Samsung Galaxy smartphone
-------- Original message --------From: Nathan Goldbaum <notifications@github.com> Date: 8/22/18  6:02 PM  (GMT-06:00) To: yt-project/yt <yt@noreply.github.com> Cc: Zach Sherman <ShermanX1991@gmail.com>, Mention <mention@noreply.github.com> Subject: Re: [yt-project/yt] [WIP] Add support for cartopy projections with geographic data (#1966) 
@zssherman how do you want to deal with upstreaming the cf_radial frontend? That should probably happen separately from this.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/yt-project/yt","title":"yt-project/yt","subtitle":"GitHub repository","main_image_url":"https://assets-cdn.github.com/images/email/message_cards/header.png","avatar_image_url":"https://assets-cdn.github.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/yt-project/yt"}},"updates":{"snippets":[{"icon":"PERSON","message":"@ngoldbaum in #1966: @zssherman how do you want to deal with upstreaming the `cf_radial` frontend? That should probably happen separately from this."}],"action":{"name":"View Pull Request","url":"https://github.com/yt-project/yt/pull/1966#issuecomment-415214707"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/yt-project/yt/pull/1966#issuecomment-415214707",
"url": "https://github.com/yt-project/yt/pull/1966#issuecomment-415214707",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
},
{
"@type": "MessageCard",
"@context": "http://schema.org/extensions",
"hideOriginalBody": "false",
"originator": "AF6C5A86-E920-430C-9C59-A73278B5EFEB",
"title": "Re: [yt-project/yt] [WIP] Add support for cartopy projections with geographic data (#1966)",
"sections": [
{
"text": "",
"activityTitle": "**Nathan Goldbaum**",
"activityImage": "https://assets-cdn.github.com/images/email/message_cards/avatar.png",
"activitySubtitle": "@ngoldbaum",
"facts": [

]
}
],
"potentialAction": [
{
"name": "Add a comment",
"@type": "ActionCard",
"inputs": [
{
"isMultiLine": true,
"@type": "TextInput",
"id": "IssueComment",
"isRequired": false
}
],
"actions": [
{
"name": "Comment",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueComment\",\n\"repositoryFullName\": \"yt-project/yt\",\n\"issueId\": 1966,\n\"IssueComment\": \"{{IssueComment.value}}\"\n}"
}
]
},
{
"name": "Close pull request",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"PullRequestClose\",\n\"repositoryFullName\": \"yt-project/yt\",\n\"pullRequestId\": 1966\n}"
},
{
"targets": [
{
"os": "default",
"uri": "https://github.com/yt-project/yt/pull/1966#issuecomment-415214707"
}
],
"@type": "OpenUri",
"name": "View on GitHub"
},
{
"name": "Unsubscribe",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"MuteNotification\",\n\"threadId\": 367242163\n}"
}
],
"themeColor": "26292E"
}
]

**Comment by munkm** [comment]
So I meant to bring up a few issues I have with this that I'd like to ask about before I take off the 'WIP' label, one of which @ngoldbaum already brought up! 
* I've been periodically merging in @zssherman 's `cf_radial` branch. Should I wait to take off the WIP label until that's merged? 
* This work is all technically independent from the `cf_radial`branch and could probably be merged separately, but I didn't want to do that unless we agree that it's the way to proceed. 
* I have the unit tests comparing the plot with `GenericImageTest` and I also do some tests with typechecking that the projection is set in our plots. The  unit test using`GenericImageTest` might be redundant but I though it might be useful to compare plots as well as the projection types. Let me know if these tests should be restructured. 
* I have added documentation, flake8 checks, and unit tests for the features I've added, but the CI won't pass because I haven't totally updated everything from the cf_radial branch. 
* I'm also happy to work with @zssherman on the plot callback annotations, but I'd like some advice on how to proceed with this PR first (i.e. how should things get merged? Should the callbacks be in this PR? Should the cf_radial get merged as is and I add the features for the callback in this PR for projections? Which branch should have the annotations WITH transform features?)

**Comment by munkm** [comment]
Also I'm sorry to hear that you're out sick @zssherman! I hope you feel better soon. 🙂

**Comment by munkm** [comment]
Ok I rebased so that the `cf-radial` branch can be submitted as a separate PR by @zssherman. I think this should separate out the projection work from the `cf-radial` frontend.

File: doc/source/visualizing/geographic_projections_and_transforms.rst [hunk 1]
```diff
@@ -0,0 +1,86 @@
+.. _geographic_projections_and_transforms:
+
+Geographic Projections and Transforms
+=====================================
+
+Geographic data that is on a sphere can be visualized by projecting that data
+onto a representation of that sphere flattened into 2d space. There exist a
+number of projection types, which can be found in the `the cartopy
+documentation <https://scitools.org.uk/cartopy/docs/v0.15/crs/projections.html>`_.
+yt now supports these projection types for geographically loaded data.
```
Comment (MatthewTurk) [inline doc/source/visualizing/geographic_projections_and_transforms.rst]: Can you change this to be:

```
`yt` now supports using these projection types, through cartopy, for geopgraphically loaded data.
```

or something that indicates more directly we're using cartopy?

File: doc/source/visualizing/geographic_projections_and_transforms.rst [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+Underlying data is assumed to have an underlying projection type of `PlateCarree`. If
```
Comment (MatthewTurk) [inline doc/source/visualizing/geographic_projections_and_transforms.rst]: Make this a link to the docs for `PlateCaree`, or at least note that `PlateCaree` means that it's lat/lon grid?

File: doc/source/visualizing/geographic_projections_and_transforms.rst [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+your data is not of this form, feel free to open an issue or file a pull
+request on the yt github page for this feature.
+
+It should be noted that
+these projections are not the same as yt's ProjectionPlot. For more information
```
Comment (MatthewTurk) [inline doc/source/visualizing/geographic_projections_and_transforms.rst]: good note!

File: doc/source/visualizing/geographic_projections_and_transforms.rst [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    bbox=bbox)
+    p = yt.SlicePlot(ds, "altitude", 'AIRDENS')
+
+If an option other than `PlateCarree` is desired, the plot projection type can
+be set manually with the `set_mpl_projection` function. This next code block
+sets the projection to a `Robinson` projection from the default `PlateCarree`.
+
+.. code-block:: python
+
+    p.set_mpl_projection('Robinson')
```
Comment (MatthewTurk) [inline doc/source/visualizing/geographic_projections_and_transforms.rst]: Is it possible to do this in the instantiation of `SlicePlot` itself, with an argument?

File: doc/source/visualizing/geographic_projections_and_transforms.rst (hunk 1 above)
Comment (munkm) [reply doc/source/visualizing/geographic_projections_and_transforms.rst]: Ah, great point! Cartopy is definitely doing all of the transformation behind-the-scenes so it should be highlighted.

**Comment by munkm** [comment]
> What do you think about making a note if the data is geographic and the projection is not set, and saying, hey, you might want to set the projection somehow.

@MatthewTurk I like this idea a lot! Which would be preferable: a log message that notifies the user about the projection that's being assumed if they don't specify it, or a prompt that explicitly asks them to choose a projection before a plot is shown?

File: doc/source/visualizing/geographic_projections_and_transforms.rst (hunk 4 above)
Comment (munkm) [reply doc/source/visualizing/geographic_projections_and_transforms.rst]: Right now I don't think so, but I could add that if it'd be useful.

**Comment by munkm** [comment]
I'm fairly certain that the geo_projections tests are failing in Python 3.6 at #5ad36a4 because of a version incompatibility between cartopy v0.16 and matplotlib v3.0.0rc2. From what I can tell `_subplot_classes` in `matplotlib.axes` was removed between v2.2.3 and v3.0.0rc2 and cartopy v0.16 is trying to call it. The most up-to-date version of cartopy has addressed this issue but there isn't a release yet.

File: yt/geometry/coordinates/geographic_coordinates.py [hunk 5]
```diff
@@ -168,11 +168,33 @@ def pixelize(self, dimension, data_source, field, bounds, size,
...
+        py = data_source["py"]
+        pdy = data_source["pdy"]
+        buff = np.zeros((size[1], size[0]), dtype="f8")
+        pixelize_cartesian(buff, px, py, pdx, pdy,
+                           data_source[field],
+                           bounds, int(antialias),
+                           period, int(periodic))
+        return buff
+
     def _ortho_pixelize(self, data_source, field, bounds, size, antialias,
```
Comment (cphyc) [inline yt/geometry/coordinates/geographic_coordinates.py]: Are you still using this one? If not, could we remove it?

File: yt/visualization/plot_window.py [hunk 6]
```diff
@@ -440,6 +451,101 @@ def set_origin(self, origin):
...
+        This will create a RoatatedPole projection with the unrotated pole
+        position at 37.5 degrees latitude and 177.5 degrees longitude by
+        passing them in as kwargs.
+
+        >>> import yt
+        >>> ds = yt.load("")
+        >>> p = yt.SlicePlot(ds, "altitude", "AIRDENS")
+        >>> p.set_mpl_projection(('RotatedPole', (), {'pole_latitude':37.5,
+        ...                       'pole_longitude':177.5}))
+        >>> p._setup_plots()
```
Comment (cphyc) [inline yt/visualization/plot_window.py]: Is there a reason why we have to call `_setup_plots()` directly after?

File: yt/visualization/tests/test_geo_projections.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+from yt.testing import ANSWER_TEST_TAG, fake_amr_ds, requires_module
+from yt.utilities.answer_testing.framework import GenericImageTest
+from yt.visualization.geo_plot_utils import transform_list, get_mpl_transform
+
+def setup():
+    """Test specific setup."""
+    from yt.config import ytcfg
+    ytcfg["yt", "__withintesting"] = "True"
+
+def compare(ds, field, idir, test_prefix, test_name, transform,
```
Comment (cphyc) [inline yt/visualization/tests/test_geo_projections.py]: This argument isn't used, is there a reason for it to be here?

File: yt/geometry/coordinates/geographic_coordinates.py (hunk 5 above)
Comment (munkm) [reply yt/geometry/coordinates/geographic_coordinates.py]: Ah good catch. I meant to remove that.

File: yt/visualization/plot_window.py (hunk 6 above)
Comment (munkm) [reply yt/visualization/plot_window.py]: The `set_mpl_projection()` function is decorated with two invalidate functions, one of which removes the axes from the plot. In order to access them again (and add annotations like coastlines)  _setup_plots() needs to be called. I think I noted this in the main documentation for this feature, but not the docstring.

File: yt/visualization/tests/test_geo_projections.py (hunk 7 above)
Comment (munkm) [reply yt/visualization/tests/test_geo_projections.py]: Woah, I totally forgot to carry that through. Thanks for pointing it out. I think it's resolved now!

File: doc/source/visualizing/geographic_projections_and_transforms.rst (hunk 4 above)
Comment (munkm) [reply doc/source/visualizing/geographic_projections_and_transforms.rst]: Ok, I've added in that functionality and updated the test and the documentation to reflect this additional functionality. We should now be able to pass in `geo_projection=`  into `SlicePlot` and the initial plot should have a projection other than `PlateCarree`.

File: doc/source/visualizing/geographic_projections_and_transforms.rst [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+be set manually either in the arguments for `SlicePlot` or 
+with the `set_mpl_projection` function. The next two code blocks show how to 
+sets the projection to a `Robinson` projection from the default `PlateCarree`
+using either method.
+
+.. code-block:: python
+
+    ds = yt.load_uniform_grid(data, sizes, 1.0, geometry=("geographic", dims),
+    bbox=bbox)
+    p = yt.SlicePlot(ds, "altitude", 'AIRDENS', geo_projection="Robinson")
```
Comment (MatthewTurk) [inline doc/source/visualizing/geographic_projections_and_transforms.rst]: I must confess I'm not *hugely* keen on this specifically being `geo_projection`, since we may want to extend it to things that aren't necessarily "geographic," but at the same time, keeping it distinct from the (yt-centric) notion of a `ProjectionPlot` is desirable.  I wonder if there's a way to balance this with something like `plot_projection` or something?

Comment (munkm) [reply doc/source/visualizing/geographic_projections_and_transforms.rst]: I agree. I struggled choosing a name for this variable so I have no strong feelings about `geo_projection`. I do think `plot_projection` works and I'd be fine with changing it to that. Another thought that I had was that if we build in support for transforms other than `PlateCarree` we are going to need to add in an argument for the transform too. We might want these variables have complementary naming conventions so their relationship is obvious. I'm not sure `plot_transform` would be totally accurate since it describes more about the data than the plot, so I might lean towards `data_transform` for that. What if we used `data_projection`?

Comment (ngoldbaum) [reply doc/source/visualizing/geographic_projections_and_transforms.rst]: So about passing in transforms and projections to the initializers - do we really need to do that? I think we already have enough arguments to the initializers and if we added arguments every time we added new functionality that list would just get longer and longer. Is it really so un-ergonomic to have a setter? Keep in mind that when we detect we're loading a geographic dataset we can just choose a default projection and use that, allowing users to customize it by calling the setter or by adding a new configuration option.

Comment (MatthewTurk) [reply doc/source/visualizing/geographic_projections_and_transforms.rst]: I'd be fine with not passing it in as long as we have an implementation for a default projection.  But we don't right now, so I'm less sure that anyone would even see this if we don't put it in the constructor.

@munkm how hard would it be to make the default projection grabbed from the coordinate handler?  I think we could have it be a property like `axis_name` or whatever, and just return a string based on the axis being viewed.

Comment (munkm) [reply doc/source/visualizing/geographic_projections_and_transforms.rst]: re: default projections -- Matt and I talked offline about this and I'm setting up something in the coordinate handler (specifically in the base handler and the geographic coordinate handler) to do this. The defaults set in the coordinate handler will be called to set `self._transform` and `self._projection` in `PlotWindow`. 

@ngoldbaum @MatthewTurk I had a thought about passing the arguments to the initializers. I'm on board with removing this for the projection argument. However, the transform argument makes more sense to send through the initializer. There's no reason that somebody would want to choose or change the data transform to be anything other than the transform that inherently describes the data. I can set a default transform in the coordinate handler, but if a user wants to override that, there's no way to do that (yet). What if I modify the existing argument in the initializer to set the transform, and not the projection?

Comment (ngoldbaum) [reply doc/source/visualizing/geographic_projections_and_transforms.rst]: Madicken can we chat in person about this issue? I'm a little confused and it will be easier to work through that in person than via github issues.

File: yt/visualization/base_plot_types.py [hunk 9]
```diff
@@ -220,10 +223,16 @@ def _init_image(self, data, cbnorm, cblinthresh, cmap, extent, aspect):
             cmap = get_brewer_cmap(cmap)
         vmin = float(self.zmin) if self.zmax is not None else None
         vmax = float(self.zmax) if self.zmax is not None else None
-        self.image = self.axes.imshow(
-            data.to_ndarray(), origin='lower', extent=extent, norm=norm,
-            vmin=vmin, vmax=vmax, aspect=aspect, cmap=cmap,
-            interpolation='nearest')
+        if self._transform:
```
Comment (munkm) [inline yt/visualization/base_plot_types.py]: I found that adding in a `transform=None` argument to imshow resulted in all of the non-transformed figures to be blank. Not passing it keeps all of the non-transformed figures behaving as normal in yt, so I needed to add this logic to ensure that no core plotting to yt changes with this PR. I've also filed an issue with matplotlib about passing a None kwarg here: https://github.com/matplotlib/matplotlib/issues/12551

File: yt/visualization/geo_plot_utils.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+def cartopy_importer(transform_name):
+    r"""Convenience function to import cartopy projection types
+    """
+
+    def _func(*args, **kwargs):
+        from yt.utilities.on_demand_imports import _cartopy as cartopy
+        return getattr(cartopy.crs, transform_name)(*args, **kwargs)
+    return _func
+
+def get_mpl_transform(mpl_proj):
```
[the remark this answers is not in the document]
Comment (munkm) [reply yt/visualization/geo_plot_utils.py]: Ok, i've modified this functionality to accept either a string, tuple or an instance of a cartopy.crs.CRS object. I think this should make it easy for users to access the transforms and projections in a few different ways. 

I'm not sure I understand what you're getting at by:
> If you really want string based lookups, then avoiding duplication would be a nice option.... getattr('cartopy.crs', transform_name).

Am I duplicating things with the way I wrote it in l18? This might be a subtle thing I'm not understanding.

[the remark this answers is not in the document]
Comment (munkm) [reply yt/visualization/geo_plot_utils.py]: Thank you so much for your review @pelson ! Your feeback really improved my PR. 

I'd like to test out passing user-defined projections (other than the ones already available in cartopy) into the stuff I wrote, but I couldn't find any examples in the cartopy documentation. Could you perhaps point me to a working example of somebody using one somewhere that I can experiment with?

File: yt/visualization/base_plot_types.py (hunk 9 above)
Comment (ImportanceOfBeingErnest) [reply yt/visualization/base_plot_types.py]: the `.transform` attribute doesn't seem to exist by default. But suppose this may be `None`, you may replace it by `ax.transData`.

File: .travis.yml [hunk 11]
```diff
@@ -56,8 +57,11 @@ install:
     pip install --upgrade pip
     pip install --upgrade wheel
     pip install --upgrade setuptools
+    if [[ -n ${CYTHON} ]]; then
```
Comment (MatthewTurk) [inline .travis.yml]: Hmm, is this related to this PR?

File: doc/source/cookbook/geographic_projections.rst [hunk 12]
```diff
@@ -0,0 +1,6 @@
+.. _cookbook-geographic_projections:
+
+Using Geographic Transforms and Projections
+-------------------------------------------
+
+.. notebook:: geographic_xforms_and_projections.ipynb
```
Comment (MatthewTurk) [inline doc/source/cookbook/geographic_projections.rst]: I looked at the rendered `geographic_xforms_and_projections.ipynb` file and noticed that it still shows the default projection to the `PlateCarree` in cell 7.  Is that still the case?

File: doc/source/visualizing/geographic_projections_and_transforms.rst [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+Geographic data that is on a sphere can be visualized by projecting that data
+onto a representation of that sphere flattened into 2d space. There exist a
+number of projection types, which can be found in the `the cartopy
+documentation <https://scitools.org.uk/cartopy/docs/latest/crs/projections.html>`_.
+With support from `cartopy <https://scitools.org.uk/cartopy/docs/latest/>`_, 
+``yt`` now supports these projection 
+types for geographically loaded data.
+Underlying data is assumed to have a transform of `PlateCarree
+<https://scitools.org.uk/cartopy/docs/latest/crs/projections.html#platecarree>`_,
+which is data on a flattened, rectangular, latitude/longitude grid. The data
```
Comment (MatthewTurk) [inline doc/source/visualizing/geographic_projections_and_transforms.rst]: Maybe just a parenthetical note that typically this is the format that data arrives in.

File: doc/source/visualizing/geographic_projections_and_transforms.rst [hunk 14]
```diff
@@ -0,0 +1,145 @@
+.. _geographic_projections_and_transforms:
+
+Geographic Projections and Transforms
```
Comment (MatthewTurk) [inline doc/source/visualizing/geographic_projections_and_transforms.rst]: We will eventually -- not as part of this PR -- make note that this is also the case for what folks in the astro simulation community might just think of as spherical data.

File: yt/geometry/coordinates/geographic_coordinates.py [hunk 15]
```diff
@@ -281,6 +288,34 @@ def image_axis_name(self):
                 ('longitude', 'altitude'),
                 ('altitude', 'latitude'))
 
+    _data_projection = None
+    @property
+    def data_projection(self):
```
Comment (MatthewTurk) [inline yt/geometry/coordinates/geographic_coordinates.py]: It might be possible to call the property on the base class, then re-assign `_data_projection[self.radial_axis]` but that's not necessary.

File: yt/visualization/plot_window.py [hunk 16]
```diff
@@ -187,6 +190,13 @@ def __init__(self, data_source, bounds, buff_size=(800,800), antialias=True,
         self.antialias = antialias
         self._axes_unit_names = None
 
+        axname = data_source.ds.coordinates.axis_name[data_source.axis]
+
+        xform = data_source.ds.coordinates.data_transform[axname]
```
Comment (MatthewTurk) [inline yt/visualization/plot_window.py]: Just a minor nit -- you mostly refer to transform all spelled out, so seeing it here abbreviated took me a second.

File: doc/source/cookbook/geographic_projections.rst (hunk 12 above)
Comment (munkm) [reply doc/source/cookbook/geographic_projections.rst]: Yes it is PlateCarree for the geographic coordinate handler. Should I change it? Or do we want to leave PlateCarree for the geographic handler and set it to Mollweide for a different handler?

File: .travis.yml (hunk 11 above)
Comment (munkm) [reply .travis.yml]: Yes it is. In some of the travis builds cartopy was installed before cython, which broke the build (e.g. https://travis-ci.org/yt-project/yt/jobs/422847999 for f38232307ff43c5605b88b578bb55012ce48f8ee ); this addressed that issue.

File: yt/visualization/base_plot_types.py (hunk 9 above)
Comment (munkm) [reply yt/visualization/base_plot_types.py]: Thanks @ImportanceOfBeingErnest ! I updated the function to set it as `ax.transData`.

File: yt/visualization/plot_window.py (hunk 16 above)
Comment (munkm) [reply yt/visualization/plot_window.py]: Ok I fixed this in b5c2d96e8016efa4d251755874cbe1e82a26f702.

File: doc/source/visualizing/geographic_projections_and_transforms.rst (hunk 13 above)
Comment (munkm) [reply doc/source/visualizing/geographic_projections_and_transforms.rst]: Addressed in b5c2d96e8016efa4d251755874cbe1e82a26f702

File: doc/source/cookbook/geographic_projections.rst (hunk 12 above)
Comment (munkm) [reply doc/source/cookbook/geographic_projections.rst]: Updated to Mollweide in b5c2d96e8016efa4d251755874cbe1e82a26f702

**Comment by munkm** [comment]
@Xarthisius pointed out that the answer test failures in python 3.6 on travis appear to be the same as those in other PRs, so I don't think they were introduced here. e.g. https://github.com/yt-project/yt/pull/2077

**Comment by MatthewTurk** [comment]
So from my reading, this is ready to review and potentially merge, right?  I think I've already "approved" it.

**Comment by munkm** [comment]
@MatthewTurk that's right! This is a working version that has all of the features discussed above. I'm happy to add more pending reviewer feedback too.

**Comment by ngoldbaum** [comment]
Yup and just because it was annoying to figure out, the pip install command that should work on macs is:

```
pip install --no-binary :all: shapely cartopy
```

**Comment by MatthewTurk** [comment]
I'm running into some issues when plotting data that doesn't cover the whole domain.

I'm using this dataset: http://ds.iris.edu/ds/products/emc-dna13/ the binary version of which is at http://ds.iris.edu/files/products/emc/data/DNA13/DNA13_percent.nc .  ( Porritt, R. W., R.M. Allen, and F. F. Pollitz. 2014. “Seismic imaging east of the Rocky Mountains with USArray.” Earth Planet. Sci. Lett. https://doi.org/10.1016/j.epsl.2013.10.034. )

Here is a gist that loads into yt:

https://gist.github.com/MatthewTurk/57360f40433817ecee4ef76bb817d970

It hangs inside cartopy's `img_transform.py` in the call to `kdtree.query`.  I inserted some print-debuggers that suggest the `target_xyz` variable has a maximum value of `inf` along the `axis=0` in the first and second locations.

I'm not entirely certain why this is happening.  Cartopy seems to be expecting some distances that are `inf`, but not *before* the kdtree query call.  I'm struggling to make this an outside-of-yt reproducible example, which is why I'm posting it here.  If I am able to fix it on our end I will post an update in the code.

**Comment by MatthewTurk** [comment]
OK, I was able to reduce it to something that doesn't use yt; now I'm trying to determine how we might avoid it in the future.

```
import cartopy.crs as ccrs
import matplotlib.pyplot as plt

ax = plt.axes(projection=ccrs.PlateCarree())
ax.coastlines()

# Save the plot by calling plt.savefig() BEFORE plt.show()
plt.savefig('coastlines.pdf')
plt.savefig('coastlines.png')

plt.show()
```

**Comment by MatthewTurk** [comment]
I submitted an issue upstream here: https://github.com/SciTools/cartopy/issues/1184

File: doc/source/visualizing/geographic_projections_and_transforms.rst [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+
+
+On ubuntu you'll need to install the following packages: ``libproj-dev``,
+``proj-data``, ``proj-bin``, and ``libgeos-dev``.
+
+Using Basic Transforms
+^^^^^^^^^^^^^^^^^^^^^^^
+
+As mentioned above, the default data transform is assumed to be of `PlateCarree
+<https://scitools.org.uk/cartopy/docs/latest/crs/projections.html#platecarree>`_,
```
Comment (brittonsmith) [inline doc/source/visualizing/geographic_projections_and_transforms.rst]: This is a very minor note. There are multiple places in this doc file with a link where the link text is "PlateCarree." This will cause sphinx to think a label has multiple definition and issue a warning. This can be fixed by making the trailing underscore into two underscores, e.g.:
```
`PlateCarree
<https://scitools.org.uk/cartopy/docs/latest/crs/projections.html#platecarree>`__
```

File: doc/source/visualizing/geographic_projections_and_transforms.rst [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+
+Geographic data that is on a sphere can be visualized by projecting that data
+onto a representation of that sphere flattened into 2d space. There exist a
+number of projection types, which can be found in the `the cartopy
+documentation <https://scitools.org.uk/cartopy/docs/latest/crs/projections.html>`_.
+With support from `cartopy <https://scitools.org.uk/cartopy/docs/latest/>`_, 
+``yt`` now supports these projection 
+types for geographically loaded data.
+Underlying data is assumed to have a transform of `PlateCarree
+<https://scitools.org.uk/cartopy/docs/latest/crs/projections.html#platecarree>`_,
```
Comment (brittonsmith) [inline doc/source/visualizing/geographic_projections_and_transforms.rst]: There's the other one.

**Comment by ngoldbaum** [comment]
@munkm along with britton's suggestion, maybe we should try testing with cartopy 0.17.0 now that that has been released. Supposedly that will fix the perf issues that Matt ran into. If updating cartopy causes issues we can back that out and go in and fix those issues in a separate PR.

**Comment by ngoldbaum** [comment]
BTW with Britton approving I feel comfortable merging this.

File: doc/source/visualizing/geographic_projections_and_transforms.rst (hunk 17 above)
Comment (munkm) [reply doc/source/visualizing/geographic_projections_and_transforms.rst]: Ah, thanks for that @brittonsmith! I really appreciate the explanation too. It should be fixed now.
